**Milestone 3**

Nama    : Rakhmat Safri Agung Pratama

Batch   : CODA-RMT-005

Notebook ini berisikan code tentang Eksplorasi Data Analisis dan Validasi Data

# A. Data Loading

* **Import Library**

In [14]:
import pandas as pd
from great_expectations.data_context import FileDataContext
context = FileDataContext.create(project_root_dir='./')

* **Data Load**

In [2]:
df = pd.read_csv('T:\CODA\Tugas\Ph 2\Week 2\p2-coda005-rmt-m3-rakhmatsapp\P2M3_Rakhmat-Safri_data_raw.csv')

# B. Exploratory Data Analysis

* **5 Data teratas**

In [8]:
df.head()

VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

* **Cek Struktur Data**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         177866 non-null  object 
 1   County                                             177861 non-null  object 
 2   City                                               177861 non-null  object 
 3   State                                              177866 non-null  object 
 4   Postal Code                                        177861 non-null  float64
 5   Model Year                                         177866 non-null  int64  
 6   Make                                               177866 non-null  object 
 7   Model                                              177866 non-null  object 
 8   Electric Vehicle Type                              177866 non-null  object

dari hasil di atas didapati bahwa data tersebut memiliki 177866 baris data, memiliki 17 kolom, dengan  tiga kolom bertipe float, empat kolom bertipe data integer, 10 kolom bertipe data objek.

* **Cek Missing Value**

In [3]:
missing_count = df.isnull().sum()
percent_missing = missing_count/len(df)
missing_value_df = pd.DataFrame({'count_missing': missing_count, 'percent_missing': percent_missing})
missing_value_df

count_missing  \
VIN (1-10)                                                     0   
County                                                         5   
City                                                           5   
State                                                          0   
Postal Code                                                    5   
Model Year                                                     0   
Make                                                           0   
Model                                                          0   
Electric Vehicle Type                                          0   
Clean Alternative Fuel Vehicle (CAFV) Eligibility              0   
Electric Range                                                 0   
Base MSRP                                                      0   
Legislative District                                         389   
DOL Vehicle ID                                                 0   
Vehicle Location                                               9   
Electric Utility                                               5   
2020 Census Tract                                              5   

                                                   percent_missing  
VIN (1-10)                                                0.000000  
County                                                    0.000028  
City                                                      0.000028  
State                                                     0.000000  
Postal Code                                               0.000028  
Model Year                                                0.000000  
Make                                                      0.000000  
Model                                                     0.000000  
Electric Vehicle Type                                     0.000000  
Clean Alternative Fuel Vehicle (CAFV) Eligibility         0.000000  
Electric Range                                            0.000000  
Base MSRP                                                 0.000000  
Legislative District                                      0.002187  
DOL Vehicle ID                                            0.000000  
Vehicle Location                                          0.000051  
Electric Utility                                          0.000028  
2020 Census Tract                                         0.000028

dari hasil di atas didapati bahwa terdapat 6 kolom yang memiliki missing value.

* **Cek Duplikat Data**

In [5]:
df.duplicated().sum()

0

dari hasil di atas tidak ditemukan data duplikat

# C. Handling Missing Values

* **Drop Missing Values**

In [11]:
df = df.dropna()

* **Cek Missing Values**

In [13]:
missing_count = df.isnull().sum()
percent_missing = missing_count/len(df)
missing_value_df = pd.DataFrame({'count_missing': missing_count, 'percent_missing': percent_missing})
missing_value_df

count_missing  \
VIN (1-10)                                                     0   
County                                                         0   
City                                                           0   
State                                                          0   
Postal Code                                                    0   
Model Year                                                     0   
Make                                                           0   
Model                                                          0   
Electric Vehicle Type                                          0   
Clean Alternative Fuel Vehicle (CAFV) Eligibility              0   
Electric Range                                                 0   
Base MSRP                                                      0   
Legislative District                                           0   
DOL Vehicle ID                                                 0   
Vehicle Location                                               0   
Electric Utility                                               0   
2020 Census Tract                                              0   

                                                   percent_missing  
VIN (1-10)                                                     0.0  
County                                                         0.0  
City                                                           0.0  
State                                                          0.0  
Postal Code                                                    0.0  
Model Year                                                     0.0  
Make                                                           0.0  
Model                                                          0.0  
Electric Vehicle Type                                          0.0  
Clean Alternative Fuel Vehicle (CAFV) Eligibility              0.0  
Electric Range                                                 0.0  
Base MSRP                                                      0.0  
Legislative District                                           0.0  
DOL Vehicle ID                                                 0.0  
Vehicle Location                                               0.0  
Electric Utility                                               0.0  
2020 Census Tract                                              0.0

dari hasil di atas tidak ditemukan missing value

# D. Data Validation

* Setup Validation

In [15]:
datasource_name = 'ev-data'
datasource = context.sources.add_pandas(datasource_name)

asset_name = 'electric-vehicle-population'
asset = datasource.add_dataframe_asset(asset_name)

batch_request = asset.build_batch_request(dataframe=df)

expectation_suite_name = 'ev_data_expectations'
context.add_or_update_expectation_suite(expectation_suite_name)

validator = context.get_validator(
    batch_request=batch_request,
    expectation_suite_name=expectation_suite_name
)

* Cek Validator Data

In [16]:
validator.head()

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

* **Validasi VIN Unique**

In [17]:
result = validator.expect_column_values_to_be_unique('VIN (1-10)')

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

* **Validasi Electric Range antara 0 - 500**

In [18]:
result = validator.expect_column_values_to_be_between(
    column='Electric Range', min_value=0, max_value=500)

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

* **Validasi State "WA"**

In [19]:
result = validator.expect_column_values_to_be_in_set('State', ['WA'])

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

* **Validasi Tipe Data**

In [20]:
result = validator.expect_column_values_to_be_in_type_list('Electric Vehicle Type', ['string', 'object'])

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

* **Validasi Kolom**

In [21]:
result = validator.expect_column_to_exist('Model')

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

* **Validasi Panjang String**

In [22]:
result = validator.expect_column_value_lengths_to_be_between('Make', min_value=2)

Calculating Metrics:   0%|          | 0/9 [00:00<?, ?it/s]

* **Validasi Jumlah Nilai**

In [24]:
result = validator.expect_column_proportion_of_unique_values_to_be_between('County', min_value=0.1)

Calculating Metrics:   0%|          | 0/7 [00:00<?, ?it/s]

* Simpan Expectation_suite

In [25]:
validator.save_expectation_suite(discard_failed_expectations=False)

* Checkpoint

In [26]:
checkpoint = context.add_or_update_checkpoint(
    name='ev_data_checkpoint',
    validator=validator,
)

checkpoint_result = checkpoint.run()

Calculating Metrics:   0%|          | 0/43 [00:00<?, ?it/s]

* Build data docs

In [27]:
context.build_data_docs()

{'local_site': 'file://t:\\CODA\\Tugas\\Ph 2\\Week 2\\p2-coda005-rmt-m3-rakhmatsapp\\gx\\uncommitted/data_docs/local_site/index.html'}